### ISSUE TITLE
Google Colab Visual RL "connection reset by peer", and PPO issues

### ISSUE THREAD

**Issue by r-liu0** [haosulab/ManiSkill#88]
When running any cell with `make_vec_env()`, such as the cell with the following code:

```
env: VecEnv = make_vec_env(
    env_id,
    num_envs,
    obs_mode=obs_mode,
    reward_mode=reward_mode,
    control_mode=control_mode,
    # specify wrappers for each individual environment e.g here we specify the
    # Continuous task wrapper and pass in the max_episode_steps parameter via the partial tool
    wrappers=[
        partial(ContinuousTaskWrapper, max_episode_steps=100)
    ]
)
```

I get the error:
```
2023-05-02 18:22:48,424 - mani_skill2 - INFO - RenderServer is running at: localhost:37067
---------------------------------------------------------------------------
ConnectionResetError                      Traceback (most recent call last)
[<ipython-input-24-49bf05d387e1>](https://localhost:8080/#) in <cell line: 44>()
     42 # create num_envs training environments, with max_episode_steps=100
     43 # instead of the default 200 to speed up training
---> 44 env: VecEnv = make_vec_env(
     45     env_id,
     46     num_envs,

5 frames
[/usr/lib/python3.10/multiprocessing/connection.py](https://localhost:8080/#) in _recv(self, size, read)
    377         remaining = size
    378         while remaining > 0:
--> 379             chunk = read(handle, remaining)
    380             n = len(chunk)
    381             if n == 0:

ConnectionResetError: [Errno 104] Connection reset by peer
```

also in the following cell:

```
from stable_baselines3 import PPO

set_random_seed(0) # set SB3's global seed to 0
rollout_steps = 3200

# create our model
policy_kwargs = dict(features_extractor_class=CustomExtractor, net_arch=[256, 128])
model = PPO("MultiInputPolicy", env, policy_kwargs=policy_kwargs, verbose=1,
    n_steps=rollout_steps // num_envs, batch_size=400,
    n_epochs=5,
    tensorboard_log="./logs",
    gamma=0.8,
    target_kl=0.2
)
```

I get the following output/error:

```
Using cuda device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
[<ipython-input-26-b39f3ce5e2d2>](https://localhost:8080/#) in <cell line: 8>()
      6 # create our model
      7 policy_kwargs = dict(features_extractor_class=CustomExtractor, net_arch=[256, 128])
----> 8 model = PPO("MultiInputPolicy", env, policy_kwargs=policy_kwargs, verbose=1,
      9     n_steps=rollout_steps // num_envs, batch_size=400,
     10     n_epochs=5,

6 frames
[/usr/local/lib/python3.10/dist-packages/stable_baselines3/ppo/ppo.py](https://localhost:8080/#) in __init__(self, policy, env, learning_rate, n_steps, batch_size, n_epochs, gamma, gae_lambda, clip_range, clip_range_vf, normalize_advantage, ent_coef, vf_coef, max_grad_norm, use_sde, sde_sample_freq, target_kl, stats_window_size, tensorboard_log, policy_kwargs, verbose, seed, device, _init_setup_model)
    102         _init_setup_model: bool = True,
    103     ):
--> 104         super().__init__(
    105             policy,
    106             env,

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/on_policy_algorithm.py](https://localhost:8080/#) in __init__(self, policy, env, learning_rate, n_steps, gamma, gae_lambda, ent_coef, vf_coef, max_grad_norm, use_sde, sde_sample_freq, stats_window_size, tensorboard_log, monitor_wrapper, policy_kwargs, verbose, seed, device, _init_setup_model, supported_action_spaces)
     76         supported_action_spaces: Optional[Tuple[spaces.Space, ...]] = None,
     77     ):
---> 78         super().__init__(
     79             policy=policy,
     80             env=env,

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/base_class.py](https://localhost:8080/#) in __init__(self, policy, env, learning_rate, policy_kwargs, stats_window_size, tensorboard_log, verbose, device, support_multi_env, monitor_wrapper, seed, use_sde, sde_sample_freq, supported_action_spaces)
    162         if env is not None:
    163             env = maybe_make_env(env, self.verbose)
--> 164             env = self._wrap_env(env, self.verbose, monitor_wrapper)
    165 
    166             self.observation_space = env.observation_space

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/base_class.py](https://localhost:8080/#) in _wrap_env(env, verbose, monitor_wrapper)
    211             if verbose >= 1:
    212                 print("Wrapping the env in a DummyVecEnv.")
--> 213             env = DummyVecEnv([lambda: env])
    214 
    215         # Make sure that dict-spaces are not nested (not supported)

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/vec_env/dummy_vec_env.py](https://localhost:8080/#) in __init__(self, env_fns)
     38         VecEnv.__init__(self, len(env_fns), env.observation_space, env.action_space)
     39         obs_space = env.observation_space
---> 40         self.keys, shapes, dtypes = obs_space_info(obs_space)
     41 
     42         self.buf_obs = OrderedDict([(k, np.zeros((self.num_envs, *tuple(shapes[k])), dtype=dtypes[k])) for k in self.keys])

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/vec_env/util.py](https://localhost:8080/#) in obs_space_info(obs_space)
     58         dtypes: a dict mapping keys to dtypes.
     59     """
---> 60     check_for_nested_spaces(obs_space)
     61     if isinstance(obs_space, spaces.Dict):
     62         assert isinstance(obs_space.spaces, OrderedDict), "Dict space must have ordered subspaces"

[/usr/local/lib/python3.10/dist-packages/stable_baselines3/common/preprocessing.py](https://localhost:8080/#) in check_for_nested_spaces(obs_space)
    223         for sub_space in sub_spaces:
    224             if isinstance(sub_space, (spaces.Dict, spaces.Tuple)):
--> 225                 raise NotImplementedError(
    226                     "Nested observation spaces are not supported (Tuple/Dict space inside Tuple/Dict space)."
    227                 )

NotImplementedError: Nested observation spaces are not supported (Tuple/Dict space inside Tuple/Dict space).
```

I'm using the following options:
```

env_id = "OpenCabinetDrawer-v1"
obs_mode = "rgbd"
control_mode = "base_pd_joint_vel_arm_pd_joint_vel"
reward_mode = "dense"
```

**Comment by megatran** [comment]
Is the vectorized env issue happening on Colab, too? 

the line `ConnectionResetError: [Errno 104] Connection reset by peer` is probably relevant to https://github.com/haosulab/ManiSkill2/issues/86 (but I was running on both the GPU server as well as a local machine with a. GPU)

**Comment by StoneT2000** [comment]
Investigating now.

**Comment by StoneT2000** [comment]
Ok so currently the issue is that when using dense rewards, we rely on the visual meshes of the handles to compute a pretty good dense reward. However, the fast vec env implementation relies on rendering everything on a separate server effectively so when computing dense rewards these visual mesh data are not available actually on a per env process.

For now you will have to just use sparse reward. Alternatively you can construct a simple dense reward it'll just train somewhat slower. I'll reply once I discuss with others about what changes we will make for these visual mesh based reward functions

**Comment by megatran** [comment]
> Ok so currently the issue is that when using dense rewards, we rely on the visual meshes of the handles to compute a pretty good dense reward. However, the fast vec env implementation relies on rendering everything on a separate server effectively so when computing dense rewards these visual mesh data are not available actually on a per env process.
> 
> For now you will have to just use sparse reward. Alternatively you can construct a simple dense reward it'll just train somewhat slower. I'll reply once I discuss with others about what changes we will make for these visual mesh based reward functions

Thanks so much for looking into this! I'm pretty new to robot learning in general and couldn't find a page on the documentation that describes what the simulator does when dense/sparse reward is selected. Most of the tutorial colab pages use `dense` rewards so I just assume it's a good choice to use.

Could you help clarify my understanding a bit?

_To me, dense reward sounds like the env is giving more finite feedback in every time step, and sparse reward sounds like a less frequent amount of reward coming from the env (perhaps every other time step or when the goal is achieved?). If my mental model is accurate, then this choice of dense vs. sparse would affect PPO training, right?_

For some more context, I'm playing around with this idea of training a PPO model that can effectively learn a task using both state-space and visual embeddings (it seems like the `CustomExtractor(BaseFeaturesExtractor)` in the sample RL notebook only processes either visual or state, so I'm modifying that).  Then I want to train the DAGGER imitation learner using the expert demonstration (coming from the mentioned PPO). The primary reason for using PPO as the expert for DAGGER is the lack of available expert data for the specific task.

Do you think the choice of sparse vs. dense rewards would affect this approach? Thank you again!

**Comment by StoneT2000** [comment]
In all ManiSkill environments naive PPO/SAC will almost never solve them from state or visual based observations when using sparse rewards. This is the same for any environment in other benchmarks. If your agent can get some success in a sparse reward environment using PPO/SAC without any kind of exploration method/intrinsic motivation method then that task is too easy in my opinion (e.g. Most implementations of some kind of Cube Lifting environment)

Sparse rewards are basically giving your agent +1 if it succeeds and 0 if it does not, meaning for each episode your reward only comes at the end. Modelling sparse reward functions is an open problem but interesting to tackle since almost all environments have easy definable sparse rewards (which are just success indicators) but no easily definable dense rewards.

There is "expert" data for each task, there are ~1k demonstration trajectories for every environment with every combo of observation space x action space. However these are generated with motion planning, which may or may not impact learning.

**Comment by xiaoninghe** [comment]
> Is the vectorized env issue happening on Colab, too?
> 
> the line `ConnectionResetError: [Errno 104] Connection reset by peer` is probably relevant to #86 (but I was running on both the GPU server as well as a local machine with a. GPU)

I got the `ConnectionResetError` when I run the [2_reinforcement_learning](https://colab.research.google.com/github/haosulab/ManiSkill2/blob/main/examples/tutorials/2_reinforcement_learning.ipynb#scrollTo=shSyf4FQPR22) notebook on this line:
```
eval_env = SubprocVecEnv([make_env(env_id, record_dir="logs/videos") for i in range(1)])
```

The problem I found was due to the `stable-baselines3` version. After changing to `stable-baselines3==1.7.0` in the first cell the problem went away. Maybe you can try if that fixes anything?

Edit: Ah, just saw the bug is fixed after migration to gymnasium in v0.5.0 
https://colab.research.google.com/github/haosulab/ManiSkill2/blob/0.5.0/examples/tutorials/2_reinforcement_learning.ipynb

**Comment by StoneT2000** [comment]
update, this is fixed now once #139 is merged!

**Comment by hantao-zhou** [comment]
Nope the error still persist at 8th December of 2023
